### PR TITLE
feat: Switch from Cap'n'proto to Bincode

### LINKED ISSUE #1599: Switch from capnproto to serde
Switch from capnproto to serde with https://github.com/bincode-org/bincode, this is because capnproto is difficult to work with while serde is not. Capnproto is also very time-consuming to work with and is error prone.

### PR DESCRIPTION
**Pull request by HarryET**
Closes #1599

### CODE REVIEW

File: compiler-core/src/compiler_cache.rs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+//     App {
+//         name: String,
+//         module: Vec<String>,
+//         parameters: Vec<Type>
+//     },
+//     Fn {
+//         arguments: Vec<Type>,
+//         _return: Type
+//     },
+// }
```
Comment (lpil) [inline compiler-core/src/compiler_cache.rs:29]: Nice 👍

File: compiler-core/src/metadata.rs [hunk 2]
```diff
@@ -1,10 +1,234 @@
...
+
+    pub fn decode(
+        id_generator: UniqueIdGenerator,
+        reader: impl BufRead + bincode::de::read::Reader,
+    ) -> Result<Module> {
+        let config = bincode::config::standard();
+        let module: compiler_cache::Module =
+            bincode::serde::decode_from_reader(reader, config).expect("bincode");
+
+        Ok(Module {
```
Comment (lpil) [inline compiler-core/src/metadata.rs:162]: You can avoid all the cloning here by consuming the module when iterating over it rather than taking reference to the values. `into_iter` will help

**Review by lpil — COMMENTED** [review]
Nice one! Left some comments inline

File: compiler-core/src/compiler_cache.rs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct TypeConstructor {
+    pub _type: Arc<Type>,
+    // TODO: convert this to an int as we only need to reconstruct type vars,
+    // not other types
+    // TODO: test
```
[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/src/compiler_cache.rs:36]: Copied from the cap'n schema, what does it mean by convert to int

[the remark this answers is not in the document]
Comment (lpil) [reply compiler-core/src/compiler_cache.rs:36]: Oh yes! It means that these do not need to be types as they are parameter definitions here. All we need to know is which ones are the same, because they are always Type::Var(Generic)

[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/src/compiler_cache.rs:36]: So we just take an integer? which represents each type?

File: compiler-core/src/build/project_compiler.rs [hunk 4]
```diff
@@ -397,7 +397,7 @@ where
     ) -> Result<(), Error> {
         for path in self.io.gleam_metadata_files(&build_dir) {
             let reader = BufReader::new(self.io.reader(&path)?);
-            let module = metadata::Metadata::decode(self.ids.clone(), reader)?;
+            let module = metadata::Metadata::decode(self.ids.clone(), reader.buffer())?;
```
Comment (lpil) [inline compiler-core/src/build/project_compiler.rs]: I think this is taking just the buffer and not the full contents. I would remove the use of a reader and read the entire file.

File: compiler-core/src/metadata.rs [hunk 5]
```diff
@@ -1,10 +1,234 @@
...
+    Result,
+};
+use std::cell::RefCell;
+use std::sync::Arc;
+use std::{collections::HashMap, io::BufRead};
+
+#[derive(Debug, Clone, Copy)]
+pub struct Metadata;
+
+impl Metadata {
```
[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/src/metadata.rs]: removed with just functions now

File: compiler-core/src/compiler_cache.rs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+//     App {
+//         name: String,
+//         module: Vec<String>,
+//         parameters: Vec<Type>
+//     },
+//     Fn {
+//         arguments: Vec<Type>,
+//         _return: Type
+//     },
+// }
```
[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/src/compiler_cache.rs]: compiler cache file gone

File: compiler-core/src/compiler_cache.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/src/compiler_cache.rs]: compiler cache file gone

File: compiler-core/src/build/package_compiler.rs [hunk 7]
```diff
@@ -106,7 +109,7 @@ where
 
         // Load the cached modules that have previously been compiled
         for module in loaded.cached.into_iter() {
-            _ = existing_modules.insert(module.name.join("/"), module.clone());
+            let _ = existing_modules.insert(module.name.join("/"), module.clone());
```
[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/src/build/package_compiler.rs]: commited suggestion

File: compiler-core/src/compiler_cache.rs [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub enum Constant {
+    Int(String),
+    Float(String),
+    String(String),
+    Tuple(Vec<Constant>),
+    BitString(Vec<BitStringSegment>),
+    List {
+        _type: Arc<Type>,
```
[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/src/compiler_cache.rs]: changed

File: compiler-core/Cargo.toml [hunk 9]
```diff
@@ -67,13 +65,11 @@ toml = "0.5.8"
 # SPDX license parsing
 spdx = "0.8.0"
 # Binary format de-serialization
-bincode = "1.3.3"
+bincode = { version = "2.0.0-rc.2", features = ["serde"] }
```
[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/Cargo.toml:68]: moved to stable version

File: compiler-core/src/metadata.rs [hunk 10]
```diff
@@ -1,10 +1,234 @@
...
+    pub fn encode(data: &Module) -> Result<Vec<u8>> {
+        let span = tracing::info_span!("metadata");
+        let _enter = span.enter();
+
+        let types = data
+            .types
+            .iter()
+            .enumerate()
```
[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/src/metadata.rs]: there is now only one call which I think I need to keep as my types in the IDE break without it

File: compiler-core/src/compiler_cache.rs [hunk 11]
```diff
@@ -0,0 +1,122 @@
+use crate::type_::{Type, ValueConstructorVariant};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct Module {
```
[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/src/compiler_cache.rs]: moved to the compiler types

File: compiler-core/src/metadata.rs [hunk 12]
```diff
@@ -1,10 +1,234 @@
...
+};
+use std::cell::RefCell;
+use std::sync::Arc;
+use std::{collections::HashMap, io::BufRead};
+
+#[derive(Debug, Clone, Copy)]
+pub struct Metadata;
+
+impl Metadata {
+    pub fn encode(data: &Module) -> Result<Vec<u8>> {
```
[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/src/metadata.rs]: tidied/simplified code mostly gone

File: compiler-core/src/metadata.rs [hunk 13]
```diff
@@ -1,10 +1,234 @@
...
+            name: data.name.clone(),
+            types,
+            values,
+            accessors,
+            types_constructors,
+            package: data.package.clone(),
+        };
+
+        let config = bincode::config::standard();
+        let buffer = bincode::serde::encode_to_vec(module, config).expect("bincode");
```
[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/src/metadata.rs]: done

File: compiler-core/src/metadata.rs [hunk 14]
```diff
@@ -1,10 +1,234 @@
...
+                    .into_iter()
+                    .map(|elm| Arc::new(Metadata::decode_type(&id_generator, &elm)))
+                    .collect::<Vec<Arc<Type>>>(),
+            },
+        }
+    }
+
+    pub fn decode(
+        id_generator: UniqueIdGenerator,
+        reader: impl BufRead + bincode::de::read::Reader,
```
[the remark this answers is not in the document]
Comment (HarryET) [reply compiler-core/src/metadata.rs]: done

File: compiler-core/src/metadata.rs (hunk 2 above)
Comment (HarryET) [reply compiler-core/src/metadata.rs]: mostly solved now with simplifications

File: compiler-core/src/metadata.rs [hunk 15]
```diff
@@ -1,10 +1,98 @@
...
+    let buffer =
+        bincode::serde::encode_to_vec(data, config).expect("failed to encode metadata (bincode)");
+
+    Ok(buffer)
+}
+
+fn undo_links(id_generator: &UniqueIdGenerator, type_var: TypeVar) -> Type {
+    match type_var {
+        TypeVar::Link { type_ } => decode_type(id_generator, &type_),
+        TypeVar::Generic { .. } => Type::Var {
```
Comment (lpil) [inline compiler-core/src/metadata.rs:28]: Let's use the constructor function

https://github.com/gleam-lang/gleam/blob/00acd0734f91a3421f19bfa6c901aa74a0b56068/compiler-core/src/type_/prelude.rs#L105

File: compiler-core/src/metadata.rs [hunk 16]
```diff
@@ -1,10 +1,98 @@
...
+            type_: Arc::new(RefCell::new(TypeVar::Generic {
+                id: id_generator.next(),
+            })),
+        },
+        TypeVar::Unbound { .. } => Type::Var {
```
Comment (lpil) [inline compiler-core/src/metadata.rs:33]: This should never happen so let's panic if there is an unbound var in the cache

File: compiler-core/src/metadata.rs [hunk 17]
```diff
@@ -1,10 +1,98 @@
...
+            type_: Arc::new(RefCell::new(TypeVar::Unbound {
+                id: id_generator.next(),
+            })),
+        },
+    }
+}
+
+fn decode_type(id_generator: &UniqueIdGenerator, type_: &Type) -> Type {
+    match (*type_).clone() {
```
Comment (lpil) [inline compiler-core/src/metadata.rs:42]: Shouldn't need to clone here

File: compiler-core/src/metadata.rs [hunk 18]
```diff
@@ -1,10 +1,98 @@
...
+        },
+        TypeVar::Unbound { .. } => Type::Var {
+            type_: Arc::new(RefCell::new(TypeVar::Unbound {
+                id: id_generator.next(),
+            })),
+        },
+    }
+}
+
+fn decode_type(id_generator: &UniqueIdGenerator, type_: &Type) -> Type {
```
Comment (lpil) [inline compiler-core/src/metadata.rs:41]: Let's make this return Arc<Type> as the other type functions do.

I think this isn't the right name- it doesn't decode a type.

File: compiler-core/src/metadata.rs [hunk 19]
```diff
@@ -1,10 +1,98 @@
...
+            args: args
+                .into_iter()
+                .map(|arg| Arc::new(decode_type(&id_generator, &arg)))
+                .collect::<Vec<Arc<Type>>>(),
+        },
+        Type::Fn { args, retrn } => Type::Fn {
+            args: args
+                .into_iter()
+                .map(|arg| Arc::new(decode_type(&id_generator, &arg)))
+                .collect::<Vec<Arc<Type>>>(),
```
Comment (lpil) [inline compiler-core/src/metadata.rs:61]: You don't need any of the type annotations in this function

File: compiler-core/src/metadata.rs [hunk 20]
```diff
@@ -1,10 +1,98 @@
...
+pub fn decode(id_generator: UniqueIdGenerator, slice: &[u8]) -> Result<Module> {
+    let config = bincode::config::standard();
+    let (module, _): (Module, usize) = bincode::serde::decode_from_slice(slice, config)
+        .expect("failed to decode metadata (bincode)");
+
+    Ok(Module {
+        types: module
+            .types
+            .into_iter()
+            .enumerate()
```
Comment (lpil) [inline compiler-core/src/metadata.rs:81]: This enumerate is unused

File: compiler-core/src/metadata.rs [hunk 21]
```diff
@@ -1,10 +1,98 @@
...
+
+pub fn decode(id_generator: UniqueIdGenerator, slice: &[u8]) -> Result<Module> {
+    let config = bincode::config::standard();
+    let (module, _): (Module, usize) = bincode::serde::decode_from_slice(slice, config)
+        .expect("failed to decode metadata (bincode)");
+
+    Ok(Module {
+        types: module
+            .types
+            .into_iter()
```
Comment (lpil) [inline compiler-core/src/metadata.rs:80]: Rather than copy to a new vector we could edit in place with iter_mut()

File: compiler-core/src/metadata.rs (hunk 15 above)
Comment (HarryET) [reply compiler-core/src/metadata.rs:28]: switched to this

File: compiler-core/src/metadata.rs (hunk 16 above)
Comment (HarryET) [reply compiler-core/src/metadata.rs:33]: panicing now using `panic!()`

File: compiler-core/src/metadata.rs (hunk 17 above)
Comment (HarryET) [reply compiler-core/src/metadata.rs:42]: that gets a "cannot move" error

File: compiler-core/src/metadata.rs (hunk 19 above)
Comment (HarryET) [reply compiler-core/src/metadata.rs:61]: removed

File: compiler-core/src/metadata.rs (hunk 20 above)
Comment (HarryET) [reply compiler-core/src/metadata.rs:81]: removed

File: compiler-core/src/build/project_compiler.rs (hunk 4 above)
Comment (HarryET) [reply compiler-core/src/build/project_compiler.rs]: updated

File: compiler-core/src/metadata.rs (hunk 18 above)
Comment (HarryET) [reply compiler-core/src/metadata.rs:33]: Returns an Arc now but need to think of a better name for what it does

File: CHANGELOG.md [hunk 22]
```diff
@@ -26,6 +26,7 @@
 - The `gleam new` command gains the `--skip-github` flag to skip creation of
   `.github/*` files.
 - Added support for deno runtime for JavaScript target.
+- Compiler Cache now uses `bincode` with `serde` for serialization instead of `cap'n'proto`
```
Comment (lpil) [inline CHANGELOG.md:29]: This isn't a user facing change so we can omit it from the changelog

File: compiler-core/src/metadata.rs (hunk 18 above)
Comment (HarryET) [reply compiler-core/src/metadata.rs]: renamed!

File: compiler-core/src/metadata.rs (hunk 17 above)
Comment (lpil) [reply compiler-core/src/metadata.rs:30]: You've changed it to use ownership in the clauses below, that works 👍

File: compiler-core/src/metadata.rs [hunk 23]
```diff
@@ -1,10 +1,71 @@
...
+        }),
+        Type::Var { type_ } => undo_links(id_generator, (*type_).clone().into_inner()),
+        Type::Tuple { elems } => Arc::new(Type::Tuple {
+            elems: elems
+                .into_iter()
+                .map(|elm| regenerate_type_ids(&id_generator, &elm))
+                .collect(),
+        }),
+    }
+}
```
Comment (lpil) [inline compiler-core/src/metadata.rs:60]: Could you use the type constructor functions I linked before for these please?

File: compiler-core/src/metadata.rs (hunk 17 above)
Comment (lpil) [reply compiler-core/src/metadata.rs:30]: That's because you're now taking ownership in the clauses below. Use a reference here and clone the individual values when constructing the new values, that way less is cloned overall.

File: compiler-core/src/metadata/tests.rs [hunk 24]
```diff
@@ -150,10 +146,10 @@ fn module_with_tuple_type() {
...
 #[test]
 fn module_with_generic_type() {
-    let t0 = type_::generic_var(0);
-    let t1 = type_::generic_var(1);
-    let t7 = type_::generic_var(7);
-    let t8 = type_::generic_var(8);
+    let t0 = generic_var(0);
+    let t1 = generic_var(1);
+    let t7 = generic_var(7);
+    let t8 = generic_var(8);
```
Comment (lpil) [inline compiler-core/src/metadata/tests.rs:152]: Undo these changes please, let's keep using the qualified version 🙏

Comment (HarryET) [reply compiler-core/src/metadata/tests.rs]: done

File: compiler-core/src/metadata.rs (hunk 23 above)
Comment (HarryET) [reply compiler-core/src/metadata.rs:59]: done for all except `App` as I couldn't fine one

File: compiler-cli/src/compile_package.rs [hunk 25]
```diff
@@ -62,8 +62,8 @@ fn load_libraries(ids: &UniqueIdGenerator, lib: &Path) -> Result<im::HashMap<Str
             continue;
         }
         for module in fs::gleam_modules_metadata_paths(path)? {
-            let reader = fs::buffered_reader(module)?;
-            let module = metadata::ModuleDecoder::new(ids.clone()).read(reader)?;
+            let slice = fs::read_bytes(module)?;
```
Comment (lpil) [inline compiler-cli/src/compile_package.rs:65]: fyi this isn't a slice.

File: compiler-core/src/metadata.rs (hunk 17 above)
Comment (lpil) [reply compiler-core/src/metadata.rs:49]: Please do not resolve comments unless they have been done!

**Comment by lpil** [comment]
Build is failing and there's still one comment to address above @HarryET

File: compiler-core/src/metadata.rs (hunk 17 above)
Comment (HarryET) [reply compiler-core/src/metadata.rs:49]: Sorry 😞 will fix this now

Comment (HarryET) [reply compiler-core/src/metadata.rs]: just cloning the individual values now

File: compiler-cli/src/compile_package.rs (hunk 25 above)
Comment (HarryET) [reply compiler-cli/src/compile_package.rs]: renamed

**Comment by lpil** [comment]
btw @HarryET , we don't use merge commits in Gleam projects so if you merge I may end up having to squash your commits into 1 to remove the merge commit.

**Comment by HarryET** [comment]
Ah okay! Sorry 😞 working on finishing this

**Comment by lpil** [comment]
Thank you. No rush!

**Comment by lpil** [comment]
Still working on this @HarryET ?

**Comment by HarryET** [comment]
Yeah I'll try get it fixed this week

**Comment by lpil** [comment]
How's it going Harry? 😁

**Comment by lpil** [comment]
Closing this for now, feel free to reopen!

**Comment by sobolevn** [comment]
I am working on an alternative PR, with credits to this one. There are multiple changes since this one was created. Please, expect it tomorrow :)
